Refactor Kafka client and messaging client factory

yutiansut opened the pull request:
- Update KafkaClient to use StreamConsumer and improve error handling
- Add default broker and group ID parameters to MessagingClientFactory
- Simplify consume method with async runtime
- Minor cleanup in TWAP and VWAP strategy placeholders
- Remove unused import in format.rs

yutiansut committed:
Refactor Kafka client and messaging client factory

- Update KafkaClient to use StreamConsumer and improve error handling
- Add default broker and group ID parameters to MessagingClientFactory
- Simplify consume method with async runtime
- Minor cleanup in TWAP and VWAP strategy placeholders
- Remove unused import in format.rs

joaquinbejar reviewed (approved):
Thank you for your contribution to the project

yutiansut committed:
Implement Heikin-Ashi trading strategy with comprehensive implementation

- Add HeikinAshiStrategy struct with candle calculation and signal generation
- Implement OrderSplitStrategy trait for order management
- Include detailed tests for Heikin-Ashi candle calculation and signal generation
- Add Side enum with PartialEq and Eq derives in orders model

yutiansut commented:
feat: Implement Heikin-Ashi technical indicator strategy

Add a new technical indicator-based trading strategy using Heikin-Ashi candlesticks.
This implementation:

- Creates a Heikin-Ashi candlestick calculation algorithm
- Implements signal generation based on consecutive bullish/bearish patterns
- Integrates with the OrderSplitStrategy trait for order execution
- Includes comprehensive unit tests for calculation and signal generation

The Heikin-Ashi strategy helps filter market noise and identify trends more
clearly by using a modified candlestick calculation that smooths price action.

Resolves #123

yutiansut commented:
running 2 tests
test strategies::technical_indicator_based::heikin_ashi::tests::test_heikin_ashi_calculation ... ok
test strategies::technical_indicator_based::heikin_ashi::tests::test_signal_generation ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/unit/mod.rs (target/debug/deps/tests-6cfd1cdc1f2c0864)

running 49 tests
test config::config_test::config_test::test_kafka_config_serialization ... ok
test config::config_test::config_test::test_config_initialization_and_serialization ... ok
test config::config_test::config_test::test_missing_env_vars ... ok
test config::config_test::config_test::test_nats_config_serialization ... ok
test config::config_test::config_test::test_rabbitmq_config_serialization ... ok
test config::config_test::config_test::test_redis_config_serialization ... ok
test models::child_orders_test::child_orders_tests::test_child_order_common_order_validation ... ok
test config::config_test::config_test::test_zeromq_config_serialization ... ok
test models::child_orders_test::child_orders_tests::test_child_order_creation ... ok
test models::child_orders_test::child_orders_tests::test_child_order_validate_empty_parent_id ... ok
test models::child_orders_test::child_orders_tests::test_child_order_validate_empty_strategy_id ... ok
test models::child_orders_test::child_orders_tests::test_child_order_validate_success ... ok
test models::format_test::orders_format_tests::test_child_order_debug_display ... ok
test models::format_test::orders_format_tests::test_debug_pretty_print ... ok
test models::format_test::orders_format_tests::test_child_order_single_line ... ok
test models::format_test::orders_format_tests::test_futures_single_line ... ok
test models::format_test::orders_format_tests::test_futures_debug_display ... ok
test models::format_test::orders_format_tests::test_display_single_line ... ok
test models::format_test::orders_format_tests::test_options_debug_display ... ok
test models::format_test::orders_format_tests::test_parent_order_single_line ... ok
test models::format_test::orders_format_tests::test_options_single_line ... ok
test models::format_test::orders_format_tests::test_order_debug_display ... ok
test models::format_test::orders_format_tests::test_order_single_line ... ok
test models::format_test::orders_format_tests::test_parent_order_debug_display ... ok
test models::format_test::orders_format_tests::test_swap_debug_display ... ok
test models::format_test::orders_format_tests::test_swap_single_line ... ok
test models::orders_test::orders_tests::test_create_cfd ... ok
test models::orders_test::orders_tests::test_create_futures ... ok
test models::orders_test::orders_tests::test_create_option_type ... ok
test models::orders_test::orders_tests::test_create_options ... ok
test models::orders_test::orders_tests::test_create_order ... ok
test models::orders_test::orders_tests::test_create_order_type ... ok
test models::orders_test::orders_tests::test_create_product_type ... ok
test models::orders_test::orders_tests::test_create_side ... ok
test models::orders_test::orders_tests::test_create_swap ... ok
test models::orders_test::orders_tests::test_create_time_in_force ... ok
test models::orders_test::orders_tests::test_order_validation ... ok
test models::orders_test::orders_tests::test_order_validation_failure ... ok
test models::orders_test::orders_tests::test_order_with_futures ... ok
test models::orders_test::orders_tests::test_order_with_options ... ok
test models::parent_orders_test::parent_orders_tests::test_create_parent_order ... ok
test models::parent_orders_test::parent_orders_tests::test_parent_order_validation_failure ... ok
test models::parent_orders_test::parent_orders_tests::test_deserialize_parent_order ... ok
test models::parent_orders_test::parent_orders_tests::test_parent_order_with_cfd ... ok
test models::parent_orders_test::parent_orders_tests::test_parent_order_with_futures ... ok
test models::parent_orders_test::parent_orders_tests::test_parent_order_validation ... ok
test models::parent_orders_test::parent_orders_tests::test_parent_order_with_options ... ok
test models::parent_orders_test::parent_orders_tests::test_parent_order_with_swap ... ok
test models::parent_orders_test::parent_orders_tests::test_serialize_parent_order ... ok

test result: ok. 49 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests strategy_execution_engine

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

joaquinbejar commented:
feel free to create new issues if you miss some functionality

yutiansut committed:
Add technical indicator-based trading strategies

- Implement MACD, RSI, Bollinger Bands, and Stochastic strategies
- Export strategy structs for easy module usage
- Expand technical indicator-based trading strategy collection

yutiansut commented:
PR: Technical Indicator Strategy Implementation and Testing
Description
This commit implements several technical indicator-based trading strategies, including:
Moving Average Crossover (MA)
Relative Strength Index (RSI)
Bollinger Bands
Stochastic Oscillator
Each strategy has been implemented with proper signal generation logic and order splitting functionality. Comprehensive test cases have been added to verify the correct behavior of each strategy.
Implementation Details
Moving Average Crossover Strategy
Implemented short-term and long-term moving average calculations
Added signal generation based on golden cross (buy) and death cross (sell)
Fixed edge cases in MA calculation and signal detection
RSI Strategy
Implemented RSI calculation with configurable periods
Added overbought/oversold thresholds for signal generation
Included proper handling of price history
Bollinger Bands Strategy
Implemented middle band (SMA), upper band, and lower band calculations
Added signal generation based on price touching the bands
Included standard deviation multiplier configuration
Stochastic Oscillator Strategy
Implemented %K and %D line calculations
Added signal generation based on line crossovers in overbought/oversold regions
Included proper handling of price, high, and low history
Testing
Each strategy includes comprehensive test cases that verify:
Proper initialization of strategy parameters
Correct handling of price data
Accurate signal generation for both buy and sell conditions
Proper order splitting based on generated signals
Edge cases and error handling
Future Work
Add more technical indicators (MACD, Ichimoku Cloud, etc.)
Implement strategy combination functionality
Add performance metrics for strategy evaluation
Improve documentation with usage examples
References
The implementation is based on established technical analysis principles as described in:
"Technical Analysis of the Financial Markets" by John J. Murphy
"New Concepts in Technical Trading Systems" by J. Welles Wilder
"Bollinger on Bollinger Bands" by John Bollinger
"Technical Analysis of Stock Trends" by Robert D. Edwards and John Magee

yutiansut commented:
@joaquinbejar

yutiansut committed:
Implement Adverse Selection Strategy with Advanced Market Microstructure Analysis

- Add comprehensive AdverseSelectionStrategy with market state detection
- Implement order flow imbalance and trade size anomaly detection
- Create advanced position management with stop loss and take profit
- Add order splitting strategy with dynamic market state adaptation
- Include extensive unit tests for strategy components
- Update Cargo.toml to include rand dependency
- Modify market microstructure module to expose strategy components
